### PR TITLE
Faster Ragged Reading of Markers

### PR DESCRIPTION
**Pull request by CSSFrancis**
### Description of the change
This fixes (some) of the bugs brought up in #168  #164 

### Progress of the PR
- [x] Change implemented (can be split into several points),
- [x] add a changelog entry in the `upcoming_changes` folder (see [`upcoming_changes/README.rst`](https://github.com/hyperspy/rosettasciio/blob/main/upcoming_changes/README.rst)),
- [x] Check formatting of the changelog entry (and eventual user guide changes) in the `docs/readthedocs.org:rosettasciio` build of this PR (link in github checks)
- [x] ready for review.

This should be much faster than previous. (For large arrays this makes ragged arrays __actually__ usable)  I wanted to start adding some benchmarks but I'm not sure the best way to do that? Jupyter notebooks don't seem like the best way to do that, maybe something similar to examples?

### Minimal example of the bug fix or the new feature
```python
import time 
import matplotlib.pyplot as plt
import hyperspy.api as hs
import numpy as np

save_times_z = []
load_times_z =[]
save_times_h = []
load_times_h =[]

num_pos = [100, 500, 1000, 2000, 4000, 10000, 40000, 100000,1000000]
load_times =[]
for i in num_pos:
    np.random.seed(42)
    data = np.array([np.random.random(size=np.random.randint(0, 20)).astype(np.float64)
                  for i in range(i)], dtype=object)
        
    s = hs.signals.BaseSignal(data)
    
    tic = time.time()
    s.save("data.zspy", overwrite=True)
    toc =time.time()
    save_times_z.append(toc-tic)
    
    tic = time.time()
    hs.load("data.zspy")
    toc =time.time()
    load_times_z.append(toc-tic)
    
    tic = time.time()
    s.save("data.hspy", overwrite=True)
    toc =time.time()
    save_times_h.append(toc-tic)
    
    tic = time.time()
    hs.load("data.hspy")
    toc =time.time()
    load_times_h.append(toc-tic)

plt.plot(num_pos, load_times_z, label="loading time (zspy)")
plt.plot(num_pos, save_times_z, label="saving time (zspy)")
plt.plot(num_pos, load_times_h, label="loading time (hspy)")
plt.plot(num_pos, save_times_h, label="saving time (hspy")
plt.xlabel("number of positions")
plt.ylabel("time in sec")
plt.legend()
```

### CODE REVIEW

File: rsciio/zspy/_api.py [hunk 1]
```diff
@@ -88,7 +90,15 @@ def __init__(self, file, signal, expg, **kwargs):
...
-        """Creates a Zarr Array object for saving ragged data"""
+        """Creates a Zarr Array object for saving ragged data
+
+        Forces the number of chunks span the array if not a dask array as
+        calculating the chunks for a ragged array is not supported.
+        """
+        if isinstance(data, da.Array):
+            chunks = chunks
+        else:
+            chunks = np.prod(data.shape)
```
Comment (ericpre) [inline rsciio/zspy/_api.py]: To remove the codeQL failure:
```suggestion
        if not isinstance(data, da.Array):
            chunks = np.prod(data.shape)
```

File: rsciio/_hierarchical.py [hunk 2]
```diff
@@ -239,7 +239,11 @@ def _read_array(group, dataset_key):
             key = "ragged_shapes"
         if key in group:
             ragged_shape = group[key]
+            ragged_shape = np.array(
+                ragged_shape
+            )  # cast to numpy array to stop multiple reads
```
Comment (ericpre) [inline rsciio/_hierarchical.py:244]: It is not obvious what is meant with the comment (at least to me!) Can you elaborate a bit more please?

File: rsciio/zspy/_api.py [hunk 3]
```diff
@@ -88,7 +90,15 @@ def __init__(self, file, signal, expg, **kwargs):
 
     @staticmethod
     def _get_object_dset(group, data, key, chunks, **kwds):
-        """Creates a Zarr Array object for saving ragged data"""
+        """Creates a Zarr Array object for saving ragged data
+
+        Forces the number of chunks span the array if not a dask array as
+        calculating the chunks for a ragged array is not supported.
```
Comment (ericpre) [inline rsciio/zspy/_api.py:96]: Link to https://github.com/hyperspy/rosettasciio/issues/168?

**Review by ericpre — COMMENTED** [review]
This looks good. When I tried, I used `len` instead of `np.prod` and it seems that it was the reason why it was still slow!

I am not sure about how useful the benchmark file would be, because we may well forget about it if there isn't something that run regularly and check how long it takes.
Maybe a test with a upper limit on the execution time would be more useful?

File: rsciio/_hierarchical.py (hunk 2 above)
Comment (CSSFrancis) [reply rsciio/_hierarchical.py]: Done!

File: rsciio/zspy/_api.py (hunk 1 above)
Comment (CSSFrancis) [reply rsciio/zspy/_api.py]: Yea I'm not sure why I wrote that like that... Maybe coding too early in the morning.

**Comment by CSSFrancis** [comment]
@ericpre I'll just remove the Jupyter notebook for now. I think the information is in this PR and 168 to recreate the tests so that should be good. 

Maybe something to come back to in time.  I think that some sort of benchmarking for the different file loaders would be helpful as it would help to identify which ones are faster/slower and help to identify if certain file readers could be made faster.

**Review by ericpre — APPROVED** [review]

**Pull request merged**